### ISSUE TITLE
Spock assertion inside if-statement doesn't work

### ISSUE THREAD

**Issue by lipnitsk** [spockframework/spock#1160]
See [this blog post](https://e.printstacktrace.blog/spock-assertion-inside-if-statement-doesnt-work-why/) for a thorough analysis of the issue.

Given test case:
```groovy
def "(2) should fail on expected == result comparison"() {
        given:
        def expected = "Hello, John!"

        when:
        def result = "Hello, Joe!"

        then:
        if (expected) {
            expected == result
        }
    }
```
Expected result: FAIL
Actual result: PASS

**Comment by Vampire** [comment]
Is this meant as bug report or feature request?

**Comment by lipnitsk** [comment]
I'm not sure whether to call it a bug, but it has definitely caused missed test failures in my experience. At the very least, the current behavior is counterintuitive and error-prone

**Comment by Vampire** [comment]
Well, it is the expected and documented behavior.
http://spockframework.org/spock/docs/1.3/all_in_one.html#_implicit_and_explicit_conditions clearly says

> Except for calls to void methods and expressions classified as interactions, all top-level expressions in these blocks are implicitly treated as conditions. To use conditions in other places, you need to designate them with Groovy’s assert keyword

Besides that, personally I mostly never trust a test I did not see fail once and for the expected reason. :-D

**Comment by lipnitsk** [comment]
You are right, it is documented, yet still error-prone. Any tips on how to enforce assertions on non-top-level conditions in a large test codebase? Is this completely out of scope for Spock and needs to be solved via CodeNarc or something similar?

**Comment by Vampire** [comment]
Well, do you have a good set of rules for verification? Just using any expression is bad too. Imagine for example `expect: result.any { it != null }`.

**Comment by winfriedgerlach** [comment]
@Vampire I suggest to close this issue, it is well-known, well-documented, intended behavior

**Comment by Vampire** [comment]
Yes and no.
Afaik @leonard84 intends to change this in the future,
so as feature request this still seems to be valid.

**Comment by winfriedgerlach** [comment]
Ah, thanks for the update!

**Comment by Vampire** [comment]
And now there is a PR that will change the behavior :-)

**Comment by winfriedgerlach** [comment]
Phew, I hope that'll go well, fingers crossed! Glad you tagged it for 3.0 👍
